url58b12a0c74_tatianaslod_daooTP_tgz-pJ8-tp1_MainTp1J8/src/server/TaskExecutorProvider.java

```diff
@@ -0,0 +1,60 @@
+package server;
+
+import daoo.ioc.TaskExecutor;
+
+import java.io.*;
+import java.util.ServiceLoader;
+
+/**
+ * Created with IntelliJ IDEA.
+ * User: Tatiana
+ * Date: 07/06/13
+ * Time: 10:52
+ * To change this template use File | Settings | File Templates.
+ */
+public class TaskExecutorProvider {
+
+    public static void printEncoders(){
+        final ServiceLoader<TaskExecutor> loader = ServiceLoader.load(TaskExecutor.class);
+        for (TaskExecutor taskExecutor : loader) {
+            System.out.println(taskExecutor.getClass().getName());
+        }
+    }
+
+    public static TaskExecutor getExecutor() throws Exception {
+        final ServiceLoader<TaskExecutor> loader = ServiceLoader.load(TaskExecutor.class);
+        for (TaskExecutor taskExecutor : loader) {
+            if(taskExecutor.getClass().getName().equals(getImplementationName())){
+                return taskExecutor;
+            }
+        }
+        throw new Exception("No such Executor found");
+    }
+
+    private static String getImplementationName() {
+        final FileReader fr;
+        try {
+            fr = new FileReader("Executor.txt");
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *fr*. The resource is referenced at lines 38, 39. The resource is closed at line 40. There are other execution paths that do not contain closure statements, for example, when BufferedReader.readLine throws an exception. Ensure *fr* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).